Add a way to configure an OpenSearch distribution for the Elasticsearch DevService

marko-bekhta opened the pull request:
Fixes #23906

This was among the things Yoann suggested to look into. While doing so, I noticed that the guide had property placeholders instead of values -- fixed that.

Also, I wanted to better understand the logic behind the used Elasticsearch/OpenSearch image versions, but couldn't figure it out 😃 so any pointers are appreciated. `elasticsearch-server.version` version that we have in POM (`7.16.3`) and is used for testing doesn't match the default we have in the config for dev services (`7.17.0`) and the ES client is `8.8.2` (while `7.10.2` for proprietary components, which is expected). As for OpenSearch, `opensearch-server.version` is `1.2.3` while there's already a 2.8.0. At least, I'd expect that the version we are testing against and the version of dev services would match...

marko-bekhta committed:
Add a way to configure an OpenSearch distribution for the Elasticsearch DevService

gsmet commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:
```diff
@@ -189,17 +187,25 @@ private DevServicesResultBuildItem.RunningDevService startElasticsearch(
+            throw new BuildException(
+                    "Dev Services for Elasticsearch detected a distribution mismatch, distribution is " + config.distribution
+                            + " but the configured distribution is " + buildItemConfig.distribution +
+                            ". Either configure a different distribution or disable Dev Services for Elasticsearch.",
```
Maybe this would be a bit better:
```suggestion
                            ". Either configure the same distribution or disable Dev Services for Elasticsearch.",
```

gsmet commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/ElasticsearchDevServicesBuildTimeConfig.java:
```diff
@@ -27,9 +27,20 @@ public class ElasticsearchDevServicesBuildTimeConfig {
+     * Defaults to the Elasticsearch distribution.
+     * <p>
+     * {@link #imageName Image name} <b>must</b> be set when
+     * the non-default ({@link Distribution#OPENSEARCH Opensearch}) distribution is used.
```
```suggestion
     * the non-default ({@link Distribution#OPENSEARCH OpenSearch}) distribution is used.
```

gsmet reviewed (dismissed):
Spotted a few minor things, otherwise that's great (but I will probably let Yoann have a look when he's back).

As for versions being a bit weird, it all boils down to us still supporting the old Elasticsearch high level client, but we should fix this soon hopefully: see https://github.com/quarkusio/quarkus/pull/31997 .

marko-bekhta commented:
Thanks for taking a look and for the link! 😃 I've addressed your suggestions, but will keep them locally for now, to not trigger more builds for now 😃 

> let Yoann have a look when he's back

yeah, I've also noticed the comments that we can only start the default backend, but couldn't think of anything obvious that would limit us from starting multiple instances, so wanted to check with Yoann on that too once he's back 😃

yrodiere commented:
> yeah, I've also noticed the comments that we can only start the default backend, but couldn't think of anything obvious that would limit us from starting multiple instances, so wanted to check with Yoann on that too once he's back smiley

See #24011, which has a few links that might explain the history of this limitation. I don't know from the top of my head why it's there. Maybe it was just the minimum viable product, and we never went beyond that.

yrodiere commented on integration-tests/hibernate-search-orm-opensearch/src/test/java/io/quarkus/it/hibernate/search/orm/opensearch/devservices/HibernateSearchElasticsearchDevServicesEnabledImplicitlyTest.java:
```diff
@@ -0,0 +1,85 @@
+        @Override
+        public Map<String, String> getConfigOverrides() {
+            return Map.of(
+                    "quarkus.elasticsearch.devservices.distribution", "opensearch",
+                    "quarkus.elasticsearch.devservices.image-name", "docker.io/opensearchproject/opensearch:2.8.0",
```
This is not really enabled "implicitly" now, is it? :) See my other comment about defaulting to opensearch based on the build item created by Hibernate Search.

yrodiere reviewed (commented):
Thanks. I agree with Guillaume's comments, and added some of my own below.

yrodiere commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:
```diff
@@ -49,6 +52,9 @@ public class DevServicesElasticsearchProcessor {
 
     private static final ContainerLocator elasticsearchContainerLocator = new ContainerLocator(DEV_SERVICE_LABEL,
             ELASTICSEARCH_PORT);
+    private static final String DEFAULT_ELASTICSEARCH_IMAGE = "docker.elastic.co/elasticsearch/elasticsearch:8.8.2";
+    private static final String DEFAULT_OPEN_SEARCH_IMAGE = "docker.io/opensearchproject/opensearch:2.9.0";
```
FYI there's a way to make sure this will always be the same image as the one in `_attributes.adoc`.

See `io.quarkus.devservices.common.ConfigureUtil#getDefaultImageNameFor` and `extensions/devservices/postgresql/src/main/resources/postgresql-devservice.properties`.

yrodiere commented on docs/src/main/asciidoc/elasticsearch-dev-services.adoc:
```diff
@@ -48,14 +49,20 @@ Note that the Elasticsearch hosts property is automatically configured with the
 quarkus.elasticsearch.devservices.image-name={elasticsearch-image}
 ----
 
+Alternatively, if an OpenSearch-based image should be used, the same property can be used:
+[source,properties,subs="attributes"]
+----
+quarkus.elasticsearch.devservices.image-name={opensearch-image}
+----
+
```
```suggestion
For exotic image names, Quarkus might be unable to determine to infer the distribution (`elastic` or `opensearch`). In these cases starting the Dev Services will fail, and you will need to configure the distribution explicitly:

[source,properties,subs="attributes"]
----
quarkus.elasticsearch.devservices.image-name=my-custom-image-with-no-clue-about-the-distribution:1.0
quarkus.elasticsearch.devservices.distribution=elasticsearch
----
```

yrodiere commented on docs/src/main/asciidoc/elasticsearch-dev-services.adoc:
```diff
@@ -48,14 +49,20 @@ Note that the Elasticsearch hosts property is automatically configured with the
 == Configuring the image
 
-Dev Services for Elasticsearch only support Elasticsearch based images, OpenSearch is not supported at the moment.
+Dev Services for Elasticsearch support distributions based on both Elasticsearch and OpenSearch images.
```
```suggestion
Dev Services for Elasticsearch support distributions based on both Elasticsearch and OpenSearch images.

When using Hibernate Search, Dev Services will default to Elasticsearch or OpenSearch based on Hibernate Search configuration.

Otherwise, Dev Services will default to Elasticsearch. To use OpenSearch, configure the distribution explicitly:
[source,properties,subs="attributes"]
----
quarkus.elasticsearch.devservices.distribution=opensearch
----

```

yrodiere commented on docs/src/main/asciidoc/elasticsearch-dev-services.adoc:
```diff
@@ -48,14 +49,20 @@ Note that the Elasticsearch hosts property is automatically configured with the
 
-If you need to use a different image than the default one you can configure it via:
-[source, properties]
+If you need to use a different Elasticsearch-based image than the default one you can configure it via:
```
```suggestion
If you need to use a different Elasticsearch or OpenSearch image than the default one you can configure it via:
```

yrodiere reviewed (commented):
Thanks! The automatic resolution of Elasticsearch/OpenSearch will be a great addition.
A few comments on the implementation below.

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:
```diff
@@ -235,6 +239,75 @@ private DevServicesResultBuildItem.RunningDevService startElasticsearch(
+        // Now let's see if we can guess it from the image:
+        if (config.imageName.isPresent()) {
+            String imageNameRepository = DockerImageName.parse(config.imageName.get()).getRepository();
+            if ("opensearchproject/opensearch".equalsIgnoreCase(imageNameRepository)) {
+                return Distribution.OPENSEARCH;
+            }
+            if ("elasticsearch/elasticsearch".equalsIgnoreCase(imageNameRepository)
+                    || "elastic/elasticsearch".equalsIgnoreCase(imageNameRepository)) {
```
mm ok, I was just thinking that if someone decides to call their image like `this-is-that-other-search-service-not-the-elasticsearch-one` 😄 we'd assume that it is elasticsearch... but I suppose those who do that are asking for trouble anyway 😄

yrodiere reviewed (approved):
LGTM, thanks!

yrodiere commented:
Ok that's a new one:

> 2023-08-14T17:15:52.0868161Z [ERROR] Tests run: 10, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.34 s <<< FAILURE! -- in io.quarkus.hibernate.orm.multiplepersistenceunits.MultiplePersistenceUnitsCdiEntityManagerTest
> 2023-08-14T17:15:52.0875976Z [ERROR] io.quarkus.hibernate.orm.multiplepersistenceunits.MultiplePersistenceUnitsCdiEntityManagerTest.usersEntityManagerInRequestNoTransaction -- Time elapsed: 6.022 s <<< FAILURE!
> 2023-08-14T17:15:52.0876971Z java.lang.AssertionError: 
> 2023-08-14T17:15:52.0877941Z 
> 2023-08-14T17:15:52.0878281Z Expecting code not to raise a throwable but caught
> 2023-08-14T17:15:52.0878708Z   "java.lang.OutOfMemoryError: Metaspace

I'll don't see how that would be caused by this PR though... I'll ask around.

gsmet commented:
I rebased to get at least one JVM build green.
